module/x/gravity/keeper/evidence_test.go

```diff
@@ -0,0 +1,132 @@
+package keeper
+
+import (
+	"encoding/hex"
+	"testing"
+	"time"
+
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/gravity-bridge/module/x/gravity/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSubmitBadSignatureEvidenceBatchExists(t *testing.T) {
```

Comment:
these are bad testing patterns... are you covering all cases? I'd suggest refactoring these as part of this PR

Reply:
This file tests:

- The cases where a validator cannot be slashed because the batch, valset, or logic call that they signed actually existed. This tests the code that sets the checkpoint when any of these items is created, as well as the code that looks for one of these checkpoints.
- The case where a validator is slashed successfully for signing a checkpoint that never existed.

It does not test:
- The case where a signature is malformed. I think we should have tests for that elsewhere, instead of testing that in every piece of code that uses our signature code.